### PR TITLE
Adds compositional pattern and implementation for SIARD 1.0 Export Module

### PR DESCRIPTION
**Pull request by chalkos**

### CODE REVIEW

File: src/main/java/com/databasepreservation/modules/siard/path/PathStrategy.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ * Namespaces should have a prefix and some parameters to build a URL
+ * Names are names of folders or files (for files it should also contain the extension)
+ *
+ * @author Bruno Ferreira <bferreira@keep.pt>
+ */
+public interface PathStrategy {
+	/**
+	 * Returns the path to the metadata.xml file
+	 */
+	public String metadataXmlFile();
```
Comment (luis100) [inline src/main/java/com/databasepreservation/modules/siard/path/PathStrategy.java]: These names should be more descriptive. Example: getMetadataXmlFilePath()
Also, what are the selected path separators? Will this work on Windows?

File: src/main/java/com/databasepreservation/modules/siard/write/FolderWriteStrategy.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+		try {
+			return Files.newOutputStream(filepath);
+		} catch (IOException e) {
+			throw new ModuleException("Error while getting the file: " + filepath.toString(),e);
+		}
+	}
+
+	@Override
+	public boolean supportsSimultaneousWriting() {
```
Comment (luis100) [inline src/main/java/com/databasepreservation/modules/siard/write/FolderWriteStrategy.java]: method name should start with "is", example: isSimultaneousWritingSupported()

File: src/main/java/com/databasepreservation/modules/siard/path/PathStrategySIARD1.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+	private static final String LOB_FILENAME = "record";
+
+	// extensions for files
+	private static final String XML_EXTENSION = "xml";
+	private static final String XSD_EXTENSION = "xsd";
+	private static final String CLOB_EXTENSION = "txt";
+	private static final String BLOB_EXTENSION = "bin";
+
+	// control characters
+	private static final String SLASH = "/";
```
Comment (luis100) [inline src/main/java/com/databasepreservation/modules/siard/path/PathStrategySIARD1.java]: Slash should be substituted by File.SEPARATOR and DOT should have a more semantic name, like FILENAME_EXTENSION_SEPARATOR.

File: src/main/java/com/databasepreservation/model/structure/DatabaseStructure.java [hunk 4]
```diff
@@ -1,13 +1,12 @@
...
-import java.util.Vector;
-
+import com.databasepreservation.utils.ListUtils;
 import org.apache.log4j.Logger;
 import org.joda.time.DateTime;
 
-import com.databasepreservation.utils.ListUtils;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Vector;
```
Comment (luis100) [inline src/main/java/com/databasepreservation/model/structure/DatabaseStructure.java]: Change all Vector to ArrayList

File: src/main/java/com/databasepreservation/modules/msAccess/in/MsAccessImportModule.java [hunk 5]
```diff
@@ -27,6 +11,17 @@
...
+import org.w3c.util.DateParser;
+
+import java.io.File;
+import java.sql.*;
+import java.text.DateFormat;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Vector;
```
Comment (luis100) [inline src/main/java/com/databasepreservation/modules/msAccess/in/MsAccessImportModule.java]: Remove all Vectors

File: src/main/java/com/databasepreservation/modules/siard/common/SIARDArchiveContainer.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
+package com.databasepreservation.modules.siard.common;
+
+import java.nio.file.Path;
+
+/**
+ * @author Bruno Ferreira <bferreira@keep.pt>
+ */
+public class SIARDArchiveContainer {
+	public enum OutputContainerType {
+		INSIDE_ARCHIVE, OUTSIDE_ARCHIVE
```
Comment (luis100) [inline src/main/java/com/databasepreservation/modules/siard/common/SIARDArchiveContainer.java]: Change to MAIN, ATTACHMENT (container)

File: src/main/java/com/databasepreservation/model/structure/DatabaseStructure.java (hunk 4 above)
Comment (chalkos) [inline src/main/java/com/databasepreservation/model/structure/DatabaseStructure.java]: https://github.com/keeps/db-preservation-toolkit/issues/44

File: src/main/java/com/databasepreservation/modules/siard/common/jaxb/siard1/CandidateKeysType.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
+//
```
Comment (luis100) [inline src/main/java/com/databasepreservation/modules/siard/common/jaxb/siard1/CandidateKeysType.java]: Remove all JAXB code and make the code generation in a maven task.

File: src/main/java/com/databasepreservation/modules/siard/common/SIARDArchiveContainer.java (hunk 6 above)
Comment (chalkos) [inline src/main/java/com/databasepreservation/modules/siard/common/SIARDArchiveContainer.java]: #45

File: src/main/java/com/databasepreservation/modules/siard/common/sql99toXSDType.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+import com.databasepreservation.model.exception.UnknownTypeException;
+import com.databasepreservation.model.structure.type.*;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * @author Bruno Ferreira <bferreira@keep.pt>
+ */
+public class sql99toXSDType {
```
Comment (luis100) [inline src/main/java/com/databasepreservation/modules/siard/common/sql99toXSDType.java]: ALways start class name capitalized.

File: src/main/java/com/databasepreservation/modules/siard/common/sql99toXSDType.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+		if (type instanceof SimpleTypeString
+				|| type instanceof SimpleTypeNumericExact
+				|| type instanceof SimpleTypeNumericApproximate
+				|| type instanceof SimpleTypeBoolean
+				|| type instanceof SimpleTypeDateTime
+				|| type instanceof SimpleTypeBinary ){
+
+			ret = convert(type.getSql99TypeName());
+
+		} else if (type instanceof UnsupportedDataType) {
```
Comment (luis100) [inline src/main/java/com/databasepreservation/modules/siard/common/sql99toXSDType.java]: throw exception when a unsupported data type is found

File: src/main/java/com/databasepreservation/modules/siard/out/content/ContentExportStrategy.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+import com.databasepreservation.model.data.Row;
+import com.databasepreservation.model.exception.ModuleException;
+import com.databasepreservation.model.structure.SchemaStructure;
+import com.databasepreservation.model.structure.TableStructure;
+
+/**
+ * @author Bruno Ferreira <bferreira@keep.pt>
+ */
+public interface ContentExportStrategy {
```
Comment (luis100) [inline src/main/java/com/databasepreservation/modules/siard/out/content/ContentExportStrategy.java]: add open and close schema, also add it on database handler

File: src/main/java/com/databasepreservation/modules/siard/common/jaxb/siard1/CandidateKeysType.java (hunk 7 above)
Comment (chalkos) [inline src/main/java/com/databasepreservation/modules/siard/common/jaxb/siard1/CandidateKeysType.java]: #46

File: src/main/java/com/databasepreservation/modules/siard/out/content/SIARD1ContentExportStrategy.java [hunk 11]
```diff
@@ -0,0 +1,424 @@
...
+
+	private final static String ENCODING = "utf-8";
+	private final static String TAB = "     ";
+	private final static String PAR = "\n";
+
+	private final ContentPathExportStrategy contentPathStrategy;
+	private final WriteStrategy writeStrategy;
+	private final SIARDArchiveContainer baseContainer;
+
+	private Set<LargeObject> LOBsToExport;
```
Comment (luis100) [inline src/main/java/com/databasepreservation/modules/siard/out/content/SIARD1ContentExportStrategy.java]: Change to List

File: src/main/java/com/databasepreservation/modules/siard/out/content/SIARD1ContentExportStrategy.java [hunk 12]
```diff
@@ -0,0 +1,424 @@
...
+
+		// reset variables
+		currentTable = null;
+		currentRowIndex = 0;
+	}
+
+	@Override
+	public void tableRow(Row row) throws ModuleException {
+		try {
+			currentWriter.append(TAB).append("<row>\n");
```
Comment (luis100) [inline src/main/java/com/databasepreservation/modules/siard/out/content/SIARD1ContentExportStrategy.java]: Make new line as a constant

File: src/main/java/com/databasepreservation/modules/siard/out/content/SIARD1ContentExportStrategy.java [hunk 13]
```diff
@@ -0,0 +1,424 @@
...
+					writeBinaryCell(cell, column, columnIndex);
+				} else if (cell instanceof SimpleCell) {
+					writeSimpleCell(cell, column, columnIndex);
+				} else if (cell instanceof ComposedCell){
+					writeComposedCell(cell, column, columnIndex);
+				}
+				// TODO add support for composed cell types
+			}
+
+			currentWriter.append(TAB).append("</row>\n");
```
Comment (luis100) [inline src/main/java/com/databasepreservation/modules/siard/out/content/SIARD1ContentExportStrategy.java]: extend stringbuilder to add prettyprint methods for adding tabs and new lines only when pretty print is enabled.

File: src/main/java/com/databasepreservation/modules/siard/common/sql99toXSDType.java (hunk 9 above)
Comment (chalkos) [inline src/main/java/com/databasepreservation/modules/siard/common/sql99toXSDType.java]: #48

File: src/main/java/com/databasepreservation/modules/siard/out/content/ContentExportStrategy.java (hunk 10 above)
Comment (chalkos) [inline src/main/java/com/databasepreservation/modules/siard/out/content/ContentExportStrategy.java]: #49

File: src/main/java/com/databasepreservation/modules/siard/common/sql99toXSDType.java (hunk 8 above)
Comment (chalkos) [inline src/main/java/com/databasepreservation/modules/siard/common/sql99toXSDType.java]: #47

File: src/main/java/com/databasepreservation/modules/msAccess/in/MsAccessImportModule.java (hunk 5 above)
Comment (chalkos) [inline src/main/java/com/databasepreservation/modules/msAccess/in/MsAccessImportModule.java]: #44

File: src/main/java/com/databasepreservation/modules/siard/out/metadata/MetadataExportStrategy.java [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+import com.databasepreservation.model.exception.ModuleException;
+import com.databasepreservation.model.structure.DatabaseStructure;
+import com.databasepreservation.modules.siard.common.SIARDArchiveContainer;
+
+/**
+ * @author Bruno Ferreira <bferreira@keep.pt>
+ */
+public interface MetadataExportStrategy {
+	void writeMetadataXML(DatabaseStructure databaseStructure, SIARDArchiveContainer container) throws ModuleException;
+	void writeMetadataXSD(DatabaseStructure databaseStructure, SIARDArchiveContainer container) throws ModuleException;
```
Comment (luis100) [inline src/main/java/com/databasepreservation/modules/siard/out/metadata/MetadataExportStrategy.java]: Add write strategy as an argument to both metadata export strategy methods. Do the same of the metadata import strategy.

File: src/main/java/com/databasepreservation/modules/siard/out/content/SIARD1ContentExportStrategy.java (hunk 11 above)
Comment (chalkos) [inline src/main/java/com/databasepreservation/modules/siard/out/content/SIARD1ContentExportStrategy.java]: #50

File: src/main/java/com/databasepreservation/modules/siard/out/metadata/SIARD1MetadataExportStrategy.java [hunk 15]
```diff
@@ -0,0 +1,786 @@
...
+		this.writeStrategy = writer;
+		this.metadataPathStrategy = metadataPathStrategy;
+	}
+
+	@Override
+	public void writeMetadataXML(DatabaseStructure dbStructure, SIARDArchiveContainer container) throws ModuleException {
+
+		JAXBContext context;
+		try {
+			context = JAXBContext.newInstance("com.databasepreservation.modules.siard.common.jaxb.siard1");
```
Comment (luis100) [inline src/main/java/com/databasepreservation/modules/siard/out/metadata/SIARD1MetadataExportStrategy.java]: Package name for SIARD should respect the SIARD namespace.

File: src/main/java/com/databasepreservation/modules/siard/out/metadata/SIARD1MetadataExportStrategy.java [hunk 16]
```diff
@@ -0,0 +1,786 @@
...
+		Reader reader = null;
+		try {
+			 reader = Files.newBufferedReader(xsdSchema, StandardCharsets.UTF_8);
+		} catch (IOException e) {
+			throw new ModuleException("Could not read from " + getClass().getResource(METADATA_XSD_RESOURCE_PATH).getPath(), e);
+		}
+
+		// read everything from reader into writer
+		try {
+			IOUtils.copy(reader, writer);
```
Comment (luis100) [inline src/main/java/com/databasepreservation/modules/siard/out/metadata/SIARD1MetadataExportStrategy.java]: Change this to use Files.copy(), reading directly from xsdSchema path: http://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#copy(java.nio.file.Path,%20java.io.OutputStream)

File: src/main/java/com/databasepreservation/modules/siard/out/content/SIARD1ContentExportStrategy.java (hunk 13 above)
Comment (chalkos) [inline src/main/java/com/databasepreservation/modules/siard/out/content/SIARD1ContentExportStrategy.java]: #51

File: src/main/java/com/databasepreservation/modules/siard/out/content/SIARD1ContentExportStrategy.java (hunk 12 above)
Comment (chalkos) [inline src/main/java/com/databasepreservation/modules/siard/out/content/SIARD1ContentExportStrategy.java]: #51

File: src/main/java/com/databasepreservation/modules/siard/out/metadata/MetadataExportStrategy.java (hunk 14 above)
Comment (chalkos) [inline src/main/java/com/databasepreservation/modules/siard/out/metadata/MetadataExportStrategy.java]: #52

File: src/main/java/com/databasepreservation/modules/siard/out/metadata/SIARD1MetadataExportStrategy.java (hunk 15 above)
Comment (chalkos) [inline src/main/java/com/databasepreservation/modules/siard/out/metadata/SIARD1MetadataExportStrategy.java]: #46

File: src/main/java/com/databasepreservation/modules/siard/out/metadata/SIARD1MetadataExportStrategy.java (hunk 16 above)
Comment (chalkos) [inline src/main/java/com/databasepreservation/modules/siard/out/metadata/SIARD1MetadataExportStrategy.java]: #53

**Pull request merged**